### PR TITLE
GH-9624 :  Adding plugin API for GetFileThumbnail method

### PR DESCRIPTION
**Pull request by pratik97**
#### Summary
This PR adds a GetFileThumbnail(fileId string) method to plugin API.

#### Ticket Link
Fixes [#9624](https://github.com/mattermost/mattermost-server/issues/9624)

### CODE REVIEW

File: app/file.go [hunk 1]
```diff
@@ -621,6 +621,30 @@ func (a *App) GetFileInfo(fileId string) (*model.FileInfo, *model.AppError) {
 	return result.Data.(*model.FileInfo), nil
 }
 
+func (a *App) GetFileThumbnail(fileId string) ([]byte, *model.AppError) {
+
+	fileInfo, err := a.GetFileInfo(fileId)
+	if err != nil {
+		mlog.Error(
+			fmt.Sprintf("File not found when getting file thumbnails, err=%v", err),
```
Comment (lieut-data) [inline app/file.go]: I'm inclined to think we should just return the error here (and below) vs. logging it: it's hard to predict how plugins might use this API, but logging errors that the plugin chooses to handle might just result in spamming the system log.

File: app/file.go [hunk 2]
```diff
@@ -621,6 +621,30 @@ func (a *App) GetFileInfo(fileId string) (*model.FileInfo, *model.AppError) {
 	return result.Data.(*model.FileInfo), nil
 }
 
+func (a *App) GetFileThumbnail(fileId string) ([]byte, *model.AppError) {
+
```
Comment (lieut-data) [inline app/file.go]: nit: can you remove this leading line?
```suggestion
```

File: app/file.go [hunk 3]
```diff
@@ -621,6 +621,30 @@ func (a *App) GetFileInfo(fileId string) (*model.FileInfo, *model.AppError) {
...
+	fileInfo, err := a.GetFileInfo(fileId)
+	if err != nil {
+		mlog.Error(
+			fmt.Sprintf("File not found when getting file thumbnails, err=%v", err),
+			mlog.String("file_id", fileId),
+		)
+		return nil, err
+	}
+
+	thumbnailPath := fileInfo.ThumbnailPath
```
Comment (lieut-data) [inline app/file.go]: If the logging is removed, you can just `return a.ReadFile(fileInfo.ThumbnailPath)` to simplify the code.

File: app/file.go [hunk 4]
```diff
@@ -621,6 +621,30 @@ func (a *App) GetFileInfo(fileId string) (*model.FileInfo, *model.AppError) {
 	return result.Data.(*model.FileInfo), nil
 }
 
+func (a *App) GetFileThumbnail(fileId string) ([]byte, *model.AppError) {
```
Comment (lieut-data) [inline app/file.go]: @jwilander, is this method strictly useful in the API, vs. a helper method that uses the existing `GetFileInfo` and `ReadFile`? I know we're on a quest to expose all the `Client4` can do, but I'm wondering how useful it is in practice to get the thumbnail without the `FileInfo`. This doesn't block the review, but just want to clarify the end goal here :)

Comment (jwilander) [reply app/file.go]: Hmm yeah this might make more sense as a helper function if there are already existing APIs that allow for this functionality

Comment (pratik97) [reply app/file.go]: @lieut-data @jwilander Do I need to make any changes?

File: plugin/api.go [hunk 5]
```diff
@@ -249,6 +249,9 @@ type API interface {
 	// Minimum server version: 5.3
 	GetFileInfo(fileId string) (*model.FileInfo, *model.AppError)
 
+	// GetFileThumbnail gets a file's thumbnail for a specific fileId.
```
Comment (hanzei) [inline plugin/api.go]: Can you document the minimum version required here?

```suggestion
	// UploadFile will upload a file to a channel using a multipart request, to be later attached to a post.
	//
	// Minimum server version: 5.6
```

**Comment by lieut-data** [comment]
@pratik97, after thinking this some more, we realized we jumped the gun on suggesting the implementation of this particular Plugin API. One of the goals for the API has been to keep the surface as small as possible, and in this particular case the `GetFileThumbnail` would be better written in the plugin code as just:
```go
	fileInfo, err := p.API.GetFileInfo(fileId)
	if err != nil {
		return nil, err
	}
	return p.API.ReadFile(fileInfo.ThumbnailPath)
```

This has the advantage that if a user needs both the `FileInfo` instance and the Thumbnail image, they won't implicitly make duplicate calls to `p.API.GetFileInfo`. Furthermore, they are empowered to use `p.API.ReadFile` for any path in any struct they find going forward.

As such, I'm inclined to close this pull request (sorry for the churn!), but I was wondering if you might instead consider extending our [mattermost-plugin-demo](https://github.com/mattermost/mattermost-plugin-demo) to use the above API methods. The demo plugin already monitors all posts on the server and duplicates them to the demo channel, so perhaps this logic could also embed a thumbnail of any image attached to such a post. 0/5, and open to discussion as such!

**Pull request closed without merging**